### PR TITLE
Chore: Use InvestmentPool state representation

### LINKED ISSUE #33: Chore: Change state representation to setup of InvestmentPool
## Status Quo
- currently the way the token states are represented in the state is a mixture of LBP & InvestmentPool
- e.g. there is an [array](https://github.com/PrimeDAO/balancer-v2-monorepo/blob/master/pkg/pool-weighted/contracts/smart/IndexPool.sol#L62) that stores all token addresses, even though we have all tokens in a [mapping](https://github.com/PrimeDAO/balancer-v2-monorepo/blob/master/pkg/pool-weighted/contracts/smart/IndexPool.sol#L51) already
- the total amount of tokens in the pool could be stored as in InvestmentPool to save space

## Acceptance Criteria
- the token state representation matches the one from the InvestmentPool

### PR DESCRIPTION
**Pull request by fabianschu**
closes #33

### CODE REVIEW

File: pkg/pool-weighted/contracts/smart/IndexPool.sol [hunk 1]
```diff
@@ -190,26 +162,31 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard {
...
-
         uint256 normalizedSum = 0;
+        bytes32 tokenState;
+
         for (uint256 i = 0; i < endWeights.length; i++) {
             uint256 endWeight = endWeights[i];
             _require(endWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
 
-            _tokenState[_tokens[i]] = _tokenState[_tokens[i]]
+            IERC20 token = tokens[i];
```
Comment (BritikovKI) [inline pkg/pool-weighted/contracts/smart/IndexPool.sol:175]: Do we really need this variable? It is used only once...

File: pkg/pool-weighted/contracts/smart/IndexPool.sol [hunk 2]
```diff
@@ -276,89 +257,96 @@ contract IndexPool is BaseWeightedPool, ReentrancyGuard {
...
+        uint256 startWeight = tokenData.decodeUint64(_START_WEIGHT_OFFSET).uncompress64();
+        uint256 endWeight = tokenData.decodeUint32(_END_WEIGHT_OFFSET).uncompress32();
+
         if (pctProgress == 0 || startWeight == endWeight) return startWeight;
         if (pctProgress >= FixedPoint.ONE) return endWeight;
 
         if (startWeight > endWeight) {
             uint256 weightDelta = pctProgress.mulDown(startWeight - endWeight);
-            return startWeight.sub(weightDelta);
+            return startWeight - weightDelta;
```
Comment (BritikovKI) [inline pkg/pool-weighted/contracts/smart/IndexPool.sol:269]: it's better to use sub and add here, it's Solidity 0.7.0, if I remember correctly Safe arithmetics is a feature of 0.8.0^

**Review by Zitzak — APPROVED** [review]

File: pkg/pool-weighted/contracts/smart/IndexPool.sol (hunk 1 above)
Comment (fabianschu) [reply pkg/pool-weighted/contracts/smart/IndexPool.sol:175]: twice :P

**Review by BritikovKI — APPROVED** [review]
Great

**Pull request merged**